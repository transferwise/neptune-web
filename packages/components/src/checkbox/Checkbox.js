import React from 'react';
import Types from 'prop-types';

import classNames from 'classnames';
import CheckboxButton from '../checkboxButton';

import { useDirection } from '../common/hooks';

const Checkbox = ({
  id,
  checked,
  required,
  disabled,
  readOnly,
  label,
  secondary,
  onChange,
  onFocus,
  onBlur,
}) => {
  const hasError = required && !disabled && !readOnly && !checked;
  const { isRTL } = useDirection();

  const classList = classNames({
    checkbox: true,
    'checkbox-lg': secondary,
    'has-error': hasError,
    disabled,
    'checkbox--rtl': isRTL,
  });

  return (
    <div id={id} className={classList}>
      {/* eslint-disable jsx-a11y/label-has-for */}
      <label>
        {label}
        {required && '*'}
        {secondary && <small>{secondary}</small>}
        <CheckboxButton
          className={classNames({ 'has-error': hasError })}
          checked={checked}
          onFocus={onFocus}
          onChange={() => onChange(!checked)}
          onBlur={onBlur}
          disabled={disabled || readOnly}
        />
      </label>
    </div>
  );
};

Checkbox.propTypes = {
  id: Types.string,
  checked: Types.bool,
  required: Types.bool,
  disabled: Types.bool,
  readOnly: Types.bool,
  label: Types.node.isRequired,
  secondary: Types.string,
  onFocus: Types.func,
  onChange: Types.func.isRequired,
  onBlur: Types.func,
};

Checkbox.defaultProps = {
  id: null,
  checked: false,
  required: false,
  disabled: false,
  readOnly: false,
  secondary: null,
  onFocus: null,
  onBlur: null,
};

export default Checkbox;
