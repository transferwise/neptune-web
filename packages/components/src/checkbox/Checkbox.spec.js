import React from 'react';
import '@testing-library/jest-dom';
import { render, fireEvent, screen } from '@testing-library/react';

import Checkbox from '.';

describe('Checkbox', () => {
  let props;

  beforeEach(() => {
    props = {
      label: <b>hello</b>,
      onChange: jest.fn(),
    };
  });

  it('renders the given label', () => {
    const { container } = render(<Checkbox {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('is enabled by default', () => {
    render(<Checkbox {...props} />);
    expect(getCheckboxContainer()).not.toHaveClass('disabled');
  });

  it('does not render secondary text', () => {
    const { container } = render(<Checkbox {...props} />);
    expect(container.querySelector('small')).not.toBeInTheDocument();
  });

  it('calls change handler with new checked value when checkbox button is clicked', () => {
    const onChange = jest.fn();

    const { rerender } = render(<Checkbox {...props} onChange={onChange} />);

    expect(onChange).not.toBeCalled();
    fireEvent.click(getLabel());
    expect(onChange).toHaveBeenCalledWith(true);

    rerender(<Checkbox {...props} onChange={onChange} checked />);

    fireEvent.click(getLabel());
    expect(onChange).toHaveBeenCalledWith(false);
  });

  it('does not call change handler on checkbox button click when disabled', () => {
    const onChange = jest.fn();

    render(<Checkbox {...props} onChange={onChange} disabled />);

    fireEvent.click(getLabel());
    expect(onChange).not.toBeCalled();
  });

  it('does not call change handler on checkbox button click when readOnly', () => {
    const onChange = jest.fn();

    render(<Checkbox {...props} onChange={onChange} readOnly />);

    fireEvent.click(getLabel());
    expect(onChange).not.toBeCalled();
  });

  it('has disabled class when the disabled prop is true', () => {
    render(<Checkbox {...props} disabled />);
    expect(getCheckboxContainer()).toHaveClass('disabled');
  });

  it('has error class and passes it to checkbox button when required and not checked', () => {
    const { rerender } = render(<Checkbox {...props} required checked />);
    expect(getCheckboxContainer()).not.toHaveClass('has-error');

    rerender(<Checkbox {...props} required />);

    expect(getCheckboxContainer()).toHaveClass('has-error');
  });

  it('ignores required if disabled or readOnly', () => {
    const { rerender } = render(<Checkbox {...props} required disabled />);
    expect(getCheckboxContainer()).not.toHaveClass('has-error');

    rerender(<Checkbox {...props} required readOnly />);
    expect(getCheckboxContainer()).not.toHaveClass('has-error');
  });

  it('has an asterisk after the label when required', () => {
    render(<Checkbox {...props} required />);

    expect(screen.getByText('*')).toBeInTheDocument();
    expect(screen.getByText('*').tagName).toBe('LABEL');
  });

  it(`doesn't passes checked to checkbox button`, () => {
    render(<Checkbox {...props} />);
    expect(getCheckbox()).not.toHaveAttribute('checked');
  });

  it('passes checked to checkbox button', () => {
    render(<Checkbox {...props} checked />);
    expect(getCheckbox()).toHaveAttribute('checked');
  });

  it('passes disabled to checkbox button', () => {
    const { rerender } = render(<Checkbox {...props} />);
    expect(getCheckbox()).not.toHaveAttribute('disabled');

    rerender(<Checkbox {...props} disabled />);
    expect(getCheckbox()).toHaveAttribute('disabled');
  });

  it('disables checkbox button if readOnly', () => {
    const { rerender } = render(<Checkbox {...props} />);

    expect(getCheckbox()).not.toHaveAttribute('disabled');

    rerender(<Checkbox {...props} readOnly />);
    expect(getCheckbox()).toHaveAttribute('disabled');
  });

  it('displays secondary text when supplied', () => {
    render(<Checkbox {...props} secondary="secondary text" />);
    expect(screen.getByText('secondary text')).toBeInTheDocument();
    expect(getCheckboxContainer()).toHaveClass('checkbox-lg');
  });

  const getCheckboxContainer = () => screen.getByText('hello').parentElement.parentElement;
  const getLabel = () => screen.getByText('hello').parentElement;
  const getCheckbox = () => screen.getByRole('checkbox');
});
