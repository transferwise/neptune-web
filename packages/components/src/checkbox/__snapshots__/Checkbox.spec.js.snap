// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders the given label 1`] = `
<div>
  <div
    class="checkbox checkbox--rtl"
  >
    <label>
      <b>
        hello
      </b>
      <span
        class="np-checkbox-button"
      >
        <input
          type="checkbox"
        />
        <span
          class="tw-checkbox-button"
        >
          <span
            class="tw-checkbox-check"
          />
        </span>
      </span>
    </label>
  </div>
</div>
`;
